### ISSUE TITLE
incomplete filter string support

### ISSUE THREAD

**Issue by hangyan** [packetcap/go-pcap#59]
using the offical cmd to test:

for udp and icmp:

<img width="1398" alt="image" src="https://github.com/user-attachments/assets/25e63931-0f2f-4a95-83e8-eb67d5403bcb">

**Comment by deitch** [comment]
Hi @hangyan 

A few things.

First, if you can, please copy text into the issue, rather than a screen capture. I know, you cannot always, but if you can, it is much more helpful.

Second, you have two example filters there. I think the first one is `udp and src host ...`, and you are saying that it works, but does icmp rather than ump packet?

Third, what is `udp and src host 192.168.227.1 dst host 192.168.227.131`. Shouldn't it be `and dst host` (adding the `and`)?

Fourth, the second filter is `icmp and src host ...`, and you are seeing an `invalid argument` error?

In any case, very happy to take PRs on this.

**Comment by hangyan** [comment]
> Hi @hangyan
> 
> A few things.
> 
> First, if you can, please copy text into the issue, rather than a screen capture. I know, you cannot always, but if you can, it is much more helpful.
> 
> Second, you have two example filters there. I think the first one is `udp and src host ...`, and you are saying that it works, but does icmp rather than ump packet?
> 
> Third, what is `udp and src host 192.168.227.1 dst host 192.168.227.131`. Shouldn't it be `and dst host` (adding the `and`)?
> 
> Fourth, the second filter is `icmp and src host ...`, and you are seeing an `invalid argument` error?
> 
> In any case, very happy to take PRs on this.

sorry , i run the tests in a vm env and i didn't configure the copy-test so screenshot is much simpler. I would like to explain the detailed errors and copy text for you.

1. `icmp and src host ...`  didn't work and it should be. unlike `tcp` and `udp`, `and` cause the invalid error, seems. this also affect `icmp6`

2. <del> `udp and src...` and `udp src...`  (only filter on src) both works, but the first one didn't filter udp packets, instead i saw tcp/icmp packets.... second one works fine </del>  please ignore this for now, cannot reproduce, maybe i was wrong.


other errors duing my tests:

1. filter on some ipv6 address failed ,  `icmp6 and dst host ::1` reports error : `unkown host`. `::1` is fine for `ping6`
2. `protocol 17` is not supported.
3.  `src 127.0.0.1 and tcp` works but `src 127.0.0.1 and icmp` didn't. error: `invalid argument`. maybe also the `and` problem.

**Comment by deitch** [comment]
> sorry , i run the tests in a vm env and i didn't configure the copy-test so screenshot is much simpler.

Yeah, sometimes you just cannot capture the text and need a screenshot. It is better than nothing.

> `icmp and src host` ... didn't work and it should be. unlike tcp and udp, seems and cause the invalid argument error. this also affect icmp6

So `tcp and src host` works, but `icmp and src host` or `icmp6 and src host` breaks? But `icmp src host` does work? 

Sounds like just a bunch of filter fixes. Would be happy to take a PR on it. Most of the comparison logic is in [compile.go](https://github.com/packetcap/go-pcap/blob/master/filter/compile.go), which creates the bpf statements. Actual compilation is in [primitive.go](https://github.com/packetcap/go-pcap/blob/master/filter/primitive.go#L101).

The real trick to this is the tests. Set up the test, and then you can slowly debug what is missing. The tests are in [compile_cases_test.go](https://github.com/packetcap/go-pcap/blob/master/filter/primitive.go#L101) and [compile_test.go](https://github.com/packetcap/go-pcap/blob/master/filter/compile_test.go)

**Comment by hangyan** [comment]
> > sorry , i run the tests in a vm env and i didn't configure the copy-test so screenshot is much simpler.
> 
> Yeah, sometimes you just cannot capture the text and need a screenshot. It is better than nothing.
> 
> > `icmp and src host` ... didn't work and it should be. unlike tcp and udp, seems and cause the invalid argument error. this also affect icmp6
> 
> So `tcp and src host` works, but `icmp and src host` or `icmp6 and src host` breaks? But `icmp src host` does work?
> 
> Sounds like just a bunch of filter fixes. Would be happy to take a PR on it. Most of the comparison logic is in [compile.go](https://github.com/packetcap/go-pcap/blob/master/filter/compile.go), which creates the bpf statements. Actual compilation is in [primitive.go](https://github.com/packetcap/go-pcap/blob/master/filter/primitive.go#L101).
> 
> The real trick to this is the tests. Set up the test, and then you can slowly debug what is missing. The tests are in [compile_cases_test.go](https://github.com/packetcap/go-pcap/blob/master/filter/primitive.go#L101) and [compile_test.go](https://github.com/packetcap/go-pcap/blob/master/filter/compile_test.go)

Yes and No. It didn't report errors, but the results is wrong:
```
root@ubuntu-5:/home/yayu/Golang/src/github.com/packetcap/go-pcap# ./dist/pcap-linux-amd64 -i lo "icmp src 192.168.227.131"
capturing from interface lo
INFO[0000] mmap buffer created at 0x7f04a109c000 with size 131072  iface=lo promiscuous=true snaplen=1600 syscalls=false timeout=0s
0: IP packet From src [192 168 227 131] to dst [192 168 227 131]
0: TCP packet From src port 34438 to dst port 8000
0: PACKET LAYER 0: Ethernet
0: PACKET LAYER 1: IPv4
0: PACKET LAYER 2: TCP
0: packet size 0, first bytes [0 0 0 0 0 0 0 0 0 0 0 0 8 0 69 0 0 60 69 77 64 0 64 6 173 22 192 168 227 131 192 168 227 131 134 134 31 64 14 105 250 64 0 0 0 0 160 2 255 215]
1: IP packet From src [192 168 227 131] to dst [192 168 227 131]
1: TCP packet From src port 34438 to dst port 8000
1: PACKET LAYER 0: Ethernet
1: PACKET LAYER 1: IPv4
1: PACKET LAYER 2: TCP
1: packet size 0, first bytes [0 0 0 0 0 0 0 0 0 0 0 0 8 0 69 0 0 60 69 77 64 0 64 6 173 22 192 168 227 131 192 168 227 131 134 134 31 64 14 105 250 64 0 0 0 0 160 2 255 215]
2: IP packet From src [192 168 227 131] to dst [192 168 227 131]
2: TCP packet From src port 8000 to dst port 34438
2: PACKET LAYER 0: Ethernet
2: PACKET LAYER 1: IPv4
2: PACKET LAYER 2: TCP
2: packet size 0, first bytes [0 0 0 0 0 0 0 0 0 0 0 0 8 0 69 0 0 40 0 0 64 0 64 6 242 119 192 168 227 131 192 168 227 131 31 64 134 134 0 0 0 0 14 105 250 65 80 20 0 0]
```

**Comment by hangyan** [comment]
seems no icmp filter instructions genrated at all.

```
Actual  : []bpf.Instruction{bpf.LoadAbsolute{Off:0xc, Size:2}, bpf.RetConstant{Val:0x40000}, bpf.RetConstant{Val:0x0}}
```

**Comment by deitch** [comment]
Yup, looks like it is missing something. Test cases should do it.

**Comment by huhong-web** [comment]
Would it be possible to add proper support for protocol-level BPF filters like `icmp` or `multicast` ?

Happy to help develop or provide more details if needed. Thanks for the great tool!

**Comment by deitch** [comment]
Sure @huhong-web it would be great to add those. Would be happy to take a PR for it. The expression tokenization happens [in expression.go](https://github.com/packetcap/go-pcap/blob/master/filter/expression.go) and the actual processing primitives happens most in [primitive.go](https://github.com/packetcap/go-pcap/blob/master/filter/primitive.go), and constants defined in [constants.go](https://github.com/packetcap/go-pcap/blob/master/filter/constants.go). It just requires proper defining and parsing and filtering, and adding the right bpf settings in [compile.go](https://github.com/packetcap/go-pcap/blob/master/filter/compile.go)
